[MRG+1] Add spherecluster related package

jasonlaska opened the pull request:
#### What does this implement/fix? Explain your changes.

Updates related projects doc with https://github.com/clara-labs/spherecluster.

jasonlaska committed:
Add spherecluster related package

jnothman commented on doc/related_projects.rst:
```diff
@@ -114,6 +114,7 @@ and tasks.
 
 - `multiisotonic <https://github.com/alexfields/multiisotonic>`_ Isotonic regression on multidimensional features.
 
+- `spherecluster <https://github.com/clara-labs/spherecluster>` _ Spherical K-means and mixture of von Mises Fisher clustering routines for the unit hypersphere.
```
that underscore needs to appear right after the backtick (no space)

jnothman replied:
"for the unit hypersphere" is awkward to me. How about "clustering onto the unit hypersphere"

jnothman commented:
This looks good, thanks! I don't intend to do a full review of your code (though if you'd like more review, you might get some attention adding it to scikit-learn-contrib), but I note that you seem to have no tests for `SphericalKMeans` except `check_estimator`, and for `VonMisesFisherMixture` you have unit tests, but no `check_estimator`. Is there a reason?

jasonlaska committed:
Address reviewer comments

I changes the text to "data on the unit hypersphere."  Hopefully that sounds less awkward (but is also precise).

jnothman commented:
Okay.

jasonlaska commented:
Thanks for reviewing.  Here are some brief answers to the qs above:

`Spherical K-means` is a near-copy of sklearn's built in K-Means, with a minor adjustment (normalizes the centroids after each maximization step).  Rather than reproducing all of K-Means tests, I just ran the basic suite.  You're right tho, it's probably worth reproducing a few tests.

In `VonMisesFisherMixture`,  I throw an exception if the data points passed in do not have unit norm, since this is a usage error and the algorithm will not work as intended.  As far as I can tell, sklearn's standard check_estimator passes in data that does not conform to this requirement and so the tests will fail.  Further, since this class is the main novel contribution of the package, I provided some unit tests;  the coverage is not complete though.

I've addressed your comments on the PR, let me know if I should change anything else!

cheers

jnothman commented:
Interesting. I'm curious because we're trying to work out how to make `check_estimator` more widely applicable... or at least so that implementers can disable some tests while retaining others.

Could you have a `normalize=True` parameter to force all points to the unit norm?

jasonlaska commented:
`normalize=True` is a good suggestion;  I'll make an issue and try to get to it soon.

Is there an open issue on `check_estimator` compatibility (or at least stating what it expects)?  I've had other problems with it in the past as it checks for all kinds of things but I couldn't find a concise spec of what's required.

jnothman commented:
#6715 is the most specifically relevant. Question is whether you could fit
this kind of data requirement into the proposed solution at #6599.

On 11 August 2016 at 10:14, Jason Laska <notifications@github.com> wrote:

> normalize=True is a good suggestion; I'll make an issue and try to get to
> it soon.
>
> Is there an open issue on check_estimator compatibility (or at least
> stating what it expects)? I've had other problems with it in the past as it
> checks for all kinds of things but I couldn't find a concise spec of what's
> required.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7172#issuecomment-239043039>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6xlkeDbjXU0nle66Ws733RgCCDX2ks5qeml-gaJpZM4JhS6k>
> .
>

jasonlaska commented:
Perhaps.  Feels like enums for the support of all input data types in a dict like that might be a tricky undertaking.  Perhaps allowing users to pass in custom data generators (or __factories__) that conform to their needs could help generalize things a bit.

nelson-liu commented:
LGTM as well.

GaelVaroquaux commented:
LGTM. Merging.
